File: 2.add-two-numbers.java
/*
 * @lc app=leetcode id=2 lang=java
 *
 * [2] Add Two Numbers
 */
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode p = l1, q = l2;
        boolean carry = false;
        while (p.next != null && q.next != null) {
            p.val += q.val;
            if (carry) {
                p.val++;
                carry = false;
            }
            if (p.val > 9) {
                carry = true;
                p.val -= 10;
            }
            p = p.next;
            q = q.next;
        }
        if (q.next != null && p.next == null) {
            p.next = q.next;
            p.val += q.val;
            if (carry) {
                p.val++;
                carry = false;
            }
            if (p.val > 9) {
                p.val -= 10;
                carry = true;
            }

        } else if (q.next == null && p.next == null) {
            p.val += q.val;
            if (carry) {
                p.val++;
            }
            if (p.val > 9) {
                p.val -= 10;
                p.next = new ListNode(1);
            }
            return l1;
        } else {
            p.val += q.val;
            if (carry) {
                p.val++;
                carry = false;
            }
            if (p.val > 9) {
                p.val -= 10;
                carry = true;
            }

        }
        p = p.next;
        while (p.next != null) {
            if (carry) {
                p.val++;
                carry = false;
            } else
                return l1;
            if (p.val > 9) {
                carry = true;
                p.val -= 10;
            }
            p = p.next;
        }
        if (carry) {
            p.val++;
        }
        if (p.val > 9) {
            p.val -= 10;
            p.next = new ListNode(1);
        }
        return l1;
    }
}

File: csharp/medium/2.add-two-numbers.cs
/*
 * @lc app=leetcode id=2 lang=csharp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        var p = l1;
        var q = l2;
        var pre = l1;
        bool carry = false;
        while(p != null && q != null) {
            p.val += q.val;
            if(carry) {
                p.val++;
                carry = false;
            }
            if(p.val > 9) {
                carry = true;
                p.val -= 10;
            }
            pre = p;
            p = p.next;
            q = q.next;
        }
        if(p == null && q == null) {
            if(carry)
                pre.next = new ListNode(1);
            return l1;
        }
        if(p == null) {
            pre.next = q;
            p = pre.next;
        }
        if(carry) {
            while(p != null) {
                p.val++;
                if(p.val > 9)
                    p.val -= 10;
                else
                    break;
                pre = p;
                p = p.next;
            }
            if(p == null)
                pre.next = new ListNode(1);
        }
        return l1;
    }
}
// @lc code=end

